File: Two Pointers/Medium/LET_no160_Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        m, n = self.getSize(headA), self.getSize(headB)
        if n > m:
            headA, headB = headB, headA
            m, n = n, m
        
        for _ in range(m - n):
            headA = headA.next
        
        while headA is not headB:
            headA, headB = headA.next, headB.next
        
        return headA
        
    def getSize(self, head):
        size = 0
        while head:
            head = head.next
            size += 1
        return size

File: Two Pointers/Easy/LET_no160_Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int m = getSize(headA), n = getSize(headB);
        if (n > m) {
            swap(headA, headB);
            swap(m, n);
        }

        for (int i = m - n; i > 0; i--)
            headA = headA->next;

        while (headA && headA != headB) {
            headA = headA->next;
            headB = headB->next;
        }

        return headA;
    }

    int getSize(ListNode *head) {
        int count = 0;
        while (head) {
            count++;
            head = head->next;
        }
        return count;
    }
};